File: unittest/EventTable_test.cc

#include <boost/test/auto_unit_test.hpp>
#include "boost/filesystem/operations.hpp"
#include "boost/filesystem/fstream.hpp"   
#include <iostream>                        
#include <fstream>
#include "eventtable.h"
using namespace soma; 
using  namespace boost;       
using namespace boost::filesystem; 
using namespace std; 

BOOST_AUTO_TEST_SUITE(EventTableTest); 

BOOST_AUTO_TEST_CASE(EventTable_create)
{

  // create a temp file
  H5::H5File::H5File * h5file 
    = new H5::H5File::H5File("EventTable_create.h5", H5F_ACC_TRUNC); 

  H5::Group grp = h5file->createGroup("testGroup");
  
  recorder::EventTable tst(grp); 
  h5file->flush(H5F_SCOPE_GLOBAL); 

  h5file->close(); 
  delete h5file; 

  int retval = system("python EventTable_test.py create");
  BOOST_CHECK_EQUAL(retval , 0); 

}

BOOST_AUTO_TEST_CASE(EventTable_append)
{
  // create a temp file
  H5::H5File::H5File * h5file 
    = new H5::H5File::H5File("EventTable_append.h5", H5F_ACC_TRUNC); 
  
  H5::Group grp = h5file->createGroup("testGroup");
  
  recorder::EventTable*  tst = new recorder::EventTable(grp); 
  int N = 1000; 
  int emod = 17; 

  uint64_t offsetval; 
  for (int i = 0; i < N; i++)
    {
      pEventPacket_t pep(new EventPacket_t); 
      pep->events.reset(new EventList_t()); 

      pep->seq = i; 
      pep->missing = false; 
      
      // now create N % 17 events
      for (int ne = 0; ne < (i % emod); ne++) {

	Event_t e; 
	e.cmd = ne;
	e.src = (i + 100) % emod; 
	for (int d = 0; d < 5; d++) {
	  e.data[d] = d + i; 
	}
	pep->events->push_back(e); 
      }
      
      
      tst->append(pep); 
      
    }
  delete tst; 
  h5file->flush(H5F_SCOPE_GLOBAL); 
  
  h5file->close(); 
  
  delete h5file; 


  int retval = system("python EventTable_test.py append");
  BOOST_CHECK_EQUAL(retval , 0); 
  

}

BOOST_AUTO_TEST_SUITE_END(); 

File: unittest/EventTable_test.py
#!/usr/bin/python
"""
This is the python file to go along with the WaveTable_test
unit test.

The first argument is the name of the test to run

"""

import tables
import sys

def typecheck(t, name, type):
    assert isinstance(t.coldescrs[name], type)
        
    

def create_test():
    t = tables.openFile('EventTable_create.h5')
    g = t.root.testGroup
    setable = t.root.testGroup.Events

    
    typecheck(setable, "seq", tables.description.UInt32Col);
    typecheck(setable, "src", tables.description.UInt8Col);
    typecheck(setable, "cmd", tables.description.UInt8Col);

    assert setable.coldescrs["data"].shape == (5,)
    
    
def append_test():
    
    f = tables.openFile('EventTable_append.h5')
    g = f.root.testGroup
    setable = g.Events

    N = 1000
    emod = 17
    pos = 0
    for i in xrange(N):
        for ne in xrange(i % 17):
            
            assert setable[pos]['seq'] == i
            assert setable[pos]['cmd'] == ne
            assert setable[pos]['src'] == (i + 100) % emod
            for d in xrange(5):
                assert setable[pos]['data'][d] == d + i
            pos += 1
                
    
    
if __name__ == "__main__":
    if sys.argv[1] == "create":
        create_test()
    elif sys.argv[1] == "append":
        append_test()
        
        
